build_templates/mkmf.rst

```diff
@@ -0,0 +1,349 @@
+mkmf
+====
+
+Introduction
+------------
+
+``mkmf`` is a tool written in perl version 5 that constructs a makefile from distributed source. ``mkmf`` typically
+produces a makefile that can compile a single executable program. But it is extensible to create a makefile for any
+purpose at all.
+
+Features of mkmf
+~~~~~~~~~~~~~~~~
+
+-  It understands dependencies in f90 (``module``\ s and ``use``), the fortran ``include`` statement, and the cpp
+   ``#include`` statement in any type of source.
+-  There are no restrictions on filenames, module names, etc.
+-  It supports the concept of overlays (where source is maintained in layers of directories with a defined precedence).
+-  It can keep track of changes to ``cpp`` flags, and knows when to recompile affected source (i.e, files containing
+   ``#ifdef``\ s that have been changed since the last invocation).
+-  It will run on any unix platform that has perl version 5 installed.
+-  It is free, and released under GPL. GFDL users can copy (or, better still, directly invoke) the file
+   ``/net/vb/public/bin/mkmf``.
+
+It can be downloaded via `GitHub <https://github.com/NOAA-GFDL/mkmf>`__. ``mkmf`` is pronounced *make-make-file* or
+*make-m-f* or even *McMuff* (Paul Kushner's suggestion).
+
+Syntax
+------
+
+The calling syntax is:
+
+``mkmf [-a abspath] [-c cppdefs] [-d] [-f] [-m makefile] [-p program] [-t template] [-v] [-w] [-x] [args]``
+
+#. ``-a abspath`` attaches the ``abspath`` at the *front* of all *relative* paths to sourcefiles.
+#. ``cppdefs`` is a list of ``cpp`` ``#define``\ s to be passed to the source files: affected object files will be
+   selectively removed if there has been a change in this state.
+#. ``-d`` is a debug flag to ``mkmf`` (much more verbose than ``-v``, but probably of use only if you are modifying
+   ``mkmf`` itself).
+#. ``-f`` is a formatting flag to restrict lines in the makefile to 256 characters. This was introduced in response to a
+   customer who wanted to edit his makefiles using ``vi``). Lines longer than that will use continuation lines as
+   needed.
+#. ``makefile`` is the name of the makefile written (default ``Makefile``).
+#. ``template`` is a file containing a list of make macros or commands written to the beginning of the makefile.
+#. ``program`` is the name of the final target (default ``a.out``)
+#. ``-v`` is a verbosity flag to ``mkmf``
+#. ``-w`` generates compile rules which use the \`wrapper' commands MPIFC and MPILD instead of FC and LD. These can then
+   be defined as the mpif90 compile scripts to ease changing between an MPI and non-MPI version.
+#. ``-x`` executes the makefile immediately.
+#. ``args`` are a list of directories and files to be searched for targets and dependencies.
```

Comment:
history: mkmf already output default namelists when i started in the project.  i did add the -w flag which uses the MPI versions of the compiler/linker for our mpi builds.  i did a little research and "in the olden days" (1980s/1990s) there was a unix version of mkmf.  balaji either took that and customized it for FMS at GFDL, or he took the idea and rewrote it himself.  since fortran wasn't the common unix language, it would be a good guess that module dependencies and default fortran namelists might be something he added.

Reply:
Since the -w flag is already documented, we're good.